packages/devtools_app/lib/src/timeline/flutter/timeline_flame_chart.dart

```diff
@@ -1,17 +1,349 @@
+
+  TimelineController _controller;
+
+  StreamSubscription _selectedEventSubscription;
+
+  int get gpuSectionStartRow => widget.data.uiEventFlow.depth;
+
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    _controller = Controllers.of(context).timeline;
+    _selectedEventSubscription?.cancel();
+    _selectedEventSubscription =
+        _controller.onSelectedTimelineEvent.listen((_) {
+      setState(() {});
+    });
+  }
+
+  @override
+  void didUpdateWidget(_FrameBasedTimelineFlameChart oldWidget) {
+    if (oldWidget.data != widget.data) {
+      setState(() {
+        _scrollControllerX.jumpTo(startingScrollPosition);
+        _scrollControllerY.jumpTo(startingScrollPosition);
+      });
+    }
+    super.didUpdateWidget(oldWidget);
+  }
+
+  @override
+  void initState() {
+    super.initState();
+
+    // TODO(kenz): improve this so we are not rebuilding on every scroll.
+    _scrollControllerX = ScrollController()
+      ..addListener(() {
+        setState(() {
+          scrollOffsetX = _scrollControllerX.offset;
+        });
+      });
+
+    _scrollControllerY = ScrollController()
+      ..addListener(() {
+        setState(() {
+          scrollOffsetY = _scrollControllerY.offset;
+        });
+      });
+  }
+
+  @override
+  void dispose() {
+    _scrollControllerX.dispose();
+    _scrollControllerY.dispose();
+    _selectedEventSubscription.cancel();
+    // TODO(kenz): dispose [_controller] here.
+    super.dispose();
+  }
+
   @override
   Widget build(BuildContext context) {
-    return Container(
-      color: Colors.black26,
-      child: const Center(
-        child: Text('TODO Flame Chart'),
-      ),
+    return LayoutBuilder(
+      builder: (context, constraints) {
+        return Scrollbar(
+          child: SingleChildScrollView(
+            controller: _scrollControllerX,
+            scrollDirection: Axis.horizontal,
+            child: Scrollbar(
+              child: SingleChildScrollView(
+                controller: _scrollControllerY,
+                scrollDirection: Axis.vertical,
+                child: _flameChartBody(constraints),
+              ),
+            ),
+          ),
+        );
+      },
+    );
+  }
+
+  Widget _flameChartBody(BoxConstraints constraints) {
+    final width = math.max(constraints.maxWidth, widget.totalStartingWidth);
+    final height = math.max(constraints.maxHeight, widget.height);
+
+    return Stack(
+      children: [
+        Container(
+          width: width,
+          height: height,
+        ),
+        ..._nodesInViewport(constraints), // pick what to show
+      ],
+    );
+  }
+
+  List<FlameChartNode> _nodesInViewport(BoxConstraints constraints) {
+    // TODO(kenz): is creating all the FlameChartNode objects expensive even if
+    // we won't add them to the view? We create all the FlameChartNode objects
+    // and place them in FlameChart rows, but we only add [nodesInViewport] to
+    // the widget tree.
```

Comment:
@DaveShuckerow thoughts?

Reply:
Creating the FlameChartNode objects is cheap as they are just simple stateless widgets so it is no different than creating any other single Dart object. Creating them on demand would be a bit better but isn't crucial.

Reply:
We're going to want to lazily build both horizontally and vertically.  We don't want to build any widgets that we don't want to show.  

ListView.builder is capable of this.

We can build one enclosing vertical view, then for each row one horizontal list view, with linked scroll controllers so that they scroll together.

https://api.flutter.dev/flutter/widgets/TrackingScrollController-class.html May be able to help with this.

```dart
ListView.builder(
  controller: normalVerticalController,
  itemBuilder: (context, row) {
    return ListView.builder(
      scrollDirection: Axis.horizontal,
      controller: _trackingHorizontalController,
       itemBuilder: (context, column) => buildCell(context, row, column)
    );
  },
);
```

This may take a bit of playing with to get it building the widgets that we place at different horizontal positions.

As an intermediate step, if you can know which rows to build already (since they're all the same height), you can use the ListView.builder to build just rows, then create a Stack for each row, containing all the items in the row.

```dart
ListView.builder(
  itemBuilder: (context, row) {
    return SingleChildScrollView(
      child: Stack(children: [every positioned child from the row that fits the current viewport]),
    );
  },
);
```

Reply:
Lets sync on this tomorrow. There are some details that make a List of lists not a good fit for this case.

Reply:
Discussed this offline. We will leave the code as is for now. Once we have a better "worst-case" scenario to play with (full timeline flame chart or CPU profile flame chart), we will see what the performance looks like.

At that point, we can
A) start by making this view a ListView of Stacks to get the lazy benefit of a ListView on the Y scrolling axis, and if that still doesn't solve performance issues we can
B) attempt to optimize in the horizontal direction with slivers and/or custom layout logic